lib/cache/reliability_manager.js

```diff
@@ -0,0 +1,164 @@
+const cluster = require('cluster');
+const path = require('path');
+const helpers = require('../helpers');
+const cm = require('../cluster_messages');
+const consts = require('../constants');
+const crypto = require('crypto');
+const _ = require('lodash');
+
+const kDbVersionReliability = 'version_reliability_manager_versions';
+const kUnreliableRootDir = '.unreliable';
+const defaultOptions = {
+    reliabilityThreshold: 0,
+    multiClient: false
+};
+
+class ReliabilityManager {
+    constructor(db, cachePath, options) {
+        this._db = db;
+        this._cachePath = cachePath;
+        this._options = options || {};
+        _.defaults(this._options, defaultOptions);
+
+        if(this._db) {
+            this._db_versionReliability = db.getCollection(kDbVersionReliability);
+            if (this._db_versionReliability == null) {
+                this._db_versionReliability = db.addCollection(kDbVersionReliability, {
+                    unique: ["versionId"],
+                    indices: ["guid", "hash"]
+                });
+            }
+        }
+
+        cm.listenFor("_updateReliabilityFactorForVersion", (data) => {
+            return this._updateReliabilityFactorForVersion(data);
+        });
+    }
+
+    /**
+     *
+     * @param {Object} params
+     * @returns {Promise<Object>}
+     * @private
+     */
+    async _updateReliabilityFactorForVersion(params) {
+        if(cluster.isWorker) {
+            return cm.send('_updateReliabilityFactorForVersion', params);
+        }
+
+        const entry = this.getEntry(params.guidStr, params.hashStr, true);
+        if(!entry.versionHash) {
+            entry.versionHash = params.versionHashStr;
+            entry.clientId = params.clientId;
```

Comment:
If the entry's clientId is set when it's first created, it looks like it will always treat the first transaction as a duplicate when multiClient is enabled

Reply:
I think you're right, will look closer at this.